### PR TITLE
Add support for dtCMatrix&dsCMatrix

### PR DESCRIPTION
**Pull request by binxiangni**
Now I just add codes for the conversion of dtCMatrix and dsCMatrix here in order to partially solve the [issue#17](https://github.com/RcppCore/RcppArmadillo/issues/17) and [issue#114](https://github.com/RcppCore/RcppArmadillo/issues/114). More conversion of sparse matrix would be added soon. Please do me a favor to check the code style or anything else important.

### CODE REVIEW

File: inst/include/RcppArmadilloAs.h [hunk 1]
```diff
@@ -87,23 +89,70 @@ namespace traits {
...
+                if(uplo == "U"){
+                    res = symmatu(res);
+                }else{
+                    res = symmatl(res);
+                }
+            }
             
             // Setting the sentinel
-            arma::access::rw(res.col_ptrs[(unsigned) dims[1] + 1]) =
+            arma::access::rw(res.col_ptrs[static_cast<unsigned>(ncol + 1)]) =
```
Comment (coatless) [inline inst/include/RcppArmadilloAs.h]: +1 for the `static_cast` usage.

File: inst/include/RcppArmadilloAs.h [hunk 2]
```diff
@@ -87,23 +89,70 @@ namespace traits {
...
+            // Creating an empty SpMat
+            arma::SpMat<T> res(static_cast<unsigned>(nrow), static_cast<unsigned>(ncol));
             
-            // Copying elements
-            std::copy(i.begin(), i.end(), arma::access::rwp(res.row_indices));
-            std::copy(p.begin(), p.end(), arma::access::rwp(res.col_ptrs));
-            std::copy(x.begin(), x.end(), arma::access::rwp(res.values));
+            // Get the type of sparse matrix
+            std::string type = Rcpp::as<std::string>(mat.slot("class"));
+            if (type == "dgCMatrix"){
```
Comment (coatless) [inline inst/include/RcppArmadilloAs.h]: Stylistically... @eddelbuettel will complain that you did not put a space between `)` and `{`.

**Comment by coatless** [comment]
Could you add unit tests that verify your assertion?

https://github.com/RcppCore/RcppArmadillo/blob/master/inst/unitTests/runit.sparse.R
https://github.com/RcppCore/RcppArmadillo/blob/master/inst/unitTests/cpp/sparse.cpp

**Comment by eddelbuettel** [comment]
Nice!  And don't worry too about indentation -- following what is in the other files in the repo is a good plan.

**Comment by binxiangni** [comment]
I am curious about the meaning of the dot and triple colons in the following codes:
```
    .setUp <- RcppArmadillo:::unit_test_setup("sparse.cpp") 
```

**Comment by coatless** [comment]
Generally, the `.` preceding a function or variable name is meant to hide it from users in a way that is similar to how dotfiles are hidden on most file systems. Basically, these functions and variables are to be considered _internal_ components of the package that shouldn't be modified.

Regarding the  `:::`, this relates to package namespaces. When you call `library(pkgname)`, you are loading all of the packages _exported_ functions and variables into the search path. If you do not, one way to access a package's function is to prefix it with the package name and two colons, e.g. `pkgname::function()`. Now, you might have components that are **not** _exported_ because they are internal components as mentioned above. To access these internal values, you can add one additional colon after the package name but before a function or variable, e.g. `RcppArmadillo:::unit_test_setup`, and viola you have access to it _outside_ the package. 

You might want to glance at [Double Colon and Triple Colon Operators](https://stat.ethz.ch/R-manual/R-devel/library/base/html/ns-dblcolon.html) help doc article for more.

File: inst/include/RcppArmadilloAs.h [hunk 3]
```diff
@@ -77,7 +77,9 @@ namespace traits {
     private:
         MATRIX mat ;
     };
-         
+    
+    // 14 June 2017
```
Comment (thirdwing) [inline inst/include/RcppArmadilloAs.h]: I think we don't have time in other place, right?

File: inst/include/RcppArmadilloAs.h [hunk 4]
```diff
@@ -87,23 +89,70 @@ namespace traits {
...
-            res.mem_resize((unsigned) x.size());
+            // Creating an empty SpMat
+            arma::SpMat<T> res(static_cast<unsigned>(nrow), static_cast<unsigned>(ncol));
             
-            // Copying elements
-            std::copy(i.begin(), i.end(), arma::access::rwp(res.row_indices));
-            std::copy(p.begin(), p.end(), arma::access::rwp(res.col_ptrs));
-            std::copy(x.begin(), x.end(), arma::access::rwp(res.values));
+            // Get the type of sparse matrix
+            std::string type = Rcpp::as<std::string>(mat.slot("class"));
```
Comment (thirdwing) [inline inst/include/RcppArmadilloAs.h]: Maybe we can print some message when converting the matrix type? What is your opinion? @coatless

**Comment by eddelbuettel** [comment]
IIRC the `sample()` code by Christian has dates in it.  I am with you that it is somewhat redundant given the commit logs.  But hey, documentation is good.

As for printing message: Hm. Maybe if and only if a `verbose` flag or option is set?

File: inst/include/RcppArmadilloAs.h (hunk 3 above)
Comment (binxiangni) [inline inst/include/RcppArmadilloAs.h]: I just followed the style of the source code. There are also dates [there](https://github.com/RcppCore/RcppArmadillo/blob/master/inst/include/RcppArmadilloAs.h#L116)

Comment (thirdwing) [inline inst/include/RcppArmadilloAs.h]: It doesn't matter. Let's focus on the unit test.

**Comment by coatless** [comment]
Perhaps a set of debug macros similar to what Rcpp does for some of the String and general conversions?

c.f.

https://github.com/RcppCore/Rcpp/blob/8f2a101789b29e572bf3abe5115d9ddb48c1b0c7/inst/include/Rcpp/String.h#L30-L41

https://github.com/RcppCore/Rcpp/blob/6f81b4684481dbd9bb554dd95e66725fc3b63a8c/inst/include/Rcpp/vector/converter.h#L50

So, we would have
```
#define DEBUG_RCPPARMADILLO_SPARSE
```
Then something print status of conversion to console with `Rprintf()` instead of `Rcpp::Rcout`?

**Comment by eddelbuettel** [comment]
Let's not get distracted.  This is Binxiang's issue ticket on something different.

**Comment by eddelbuettel** [comment]
@binxiangni The pull request looks really good.  Two more suggestions / requests

1)  Tests would be excellent, and if you look at the existing directory it should become clear how to add them.  If it is not, well then we can merge this as is and open a new branch just for tests.  They are important and a key part of how we do things.

2)  ChangeLog entries would be nice.  That gets autoformatted if you use Emacs, it is pretty easy to do by hand as well.  Date, name and email with two spaces, then a bullet point for each changed file with a summary.

Can we add that?

**Comment by binxiangni** [comment]
@eddelbuettel Sorry for delay. I'll do that ASAP.  Please do not merge for now. I understand the importance of test.

As I mentioned before, my broken laptop quite let me down. Yesterday I tried to use Ubuntu Server in the digitalocean as KK suggests and set up a GUI for further working environment, but GUI didn't work. I'll go to a local dealer to get a new laptop today. After that, I can go back to the right track. Apologize again!

**Comment by eddelbuettel** [comment]
No worry, you are doing great.  We could do it as two PRs as well but may as well have you expand this one.

**Comment by eddelbuettel** [comment]
Oh, and sorry about the hardware troubles.

@coatless Any way a student can access an proper and current dev env on campus?

**Comment by binxiangni** [comment]
The local dealer also tells me that I can get the laptop after 10 days. :sob: 
Now I try to install Rstudio in the ubuntu server in the digitalocean. There is a step not making sense for me 

> RStudio can be access through port 8787.

Do you have any idea?

**Comment by eddelbuettel** [comment]
Sure. Standard RStudio _Server_ use via a webbrowser.  See eg https://support.rstudio.com/hc/en-us/articles/200552306-Accessing-RStudio-Server-Open-Source

Really no resources on campus?

**Comment by binxiangni** [comment]
I can access the computer in the computer lab, but have no admin right to install package like Armadillo. But for now, I have an old laptop from my friend and at least it can run the `build check` in the RStudio. Previously, I think I have to install C++ Armadillo library. Maybe I am wrong. Only if the build check runs and no error occurs, that means it passes the test, right?

**Comment by eddelbuettel** [comment]
Also, you _could_ just use GitHub and Travis CI.  If you clone RcppArmadillo you have a working `.travis.yml`.  So you [just need to turn on Travis CI](https://docs.travis-ci.com/user/getting-started/) for your account as everything else is already there.  (For what it is worth, the first Google hit [was this tutorial](https://github.com/mbonaci/mbo-storm/wiki/Integrate-Travis-CI-with-your-GitHub-repo).)

**Comment by coatless** [comment]
Just to give a heads up, I've put in a request to get @binxiangni access to a linux machine on campus. I'll let you know what happens. (It likely won't be until Monday since it's almost 5 o'clock.)

**Comment by eddelbuettel** [comment]
Thanks for adding tests. Looks like one failed ... but I am about to cook dinner so can't check now.

I am sure we'll get it sorted out.

**Comment by binxiangni** [comment]
I find where the bug is. I'm working on it.

File: inst/unitTests/cpp/sparse.cpp [hunk 5]
```diff
@@ -77,3 +77,41 @@ Rcpp::List sparseList(Rcpp::List l) {
...
 }
+
+// [[Rcpp::export]]
+arma::sp_mat dtc2dgc(SEXP S){
+    S4 mat = S;
+    IntegerVector dims = mat.slot("Dim");
+    int nrow = dims[0];
+    int ncol = dims[1];
+    
+    // Creating an empty SpMat
```
Comment (thirdwing) [inline inst/unitTests/cpp/sparse.cpp]: I think what we want to test is `arma::sp_mat res = Rcpp::as<arma::sp_mat>(S);`

The converting code has been added into the `as` implementation and we want to test.

**Comment by binxiangni** [comment]
I did unit test locally and it shows `R CMD check succeeded`, which I guess means that the unit tests pass. But there is a warning: `Problems with news in 'inst/NEWS.Rd':`. Anyone can give me a hint?

**Comment by eddelbuettel** [comment]
> `Problems with news in 'inst/NEWS.Rd'`

That usually  means a closing `}` is missing or in the wrong.  And as you didn't touch the file, not your problem :)

I can take a look in a bit.  Just in a from a long run...

**Comment by eddelbuettel** [comment]
Passes with flying colors for me too!

**Pull request merged**